### ISSUE TITLE
bot-detection plugin memory leak

### ISSUE THREAD

**Issue by ms2008** [Kong/kong#3565]
### Summary

kong alloc memory is continuously increasing.

### Steps To Reproduce

1. add a api
   ```sh
   curl -i -X POST \
     --url http://localhost:8001/apis/ \
     --data 'name=test-api' \
     --data 'uris=/foo' \
     --data 'upstream_url=http://127.0.0.1'
   ```

2. add only bot-detection plugin with default config
   ```sh
   curl -i -X POST \
     --url http://localhost:8001/plugins/ \
     --data 'name=bot-detection'
   ```

3. use ab to test the api above
   ```sh
   ab -k -c 1 -n 1000000 http://localhost:8000/foo
   ```
  
4. capture gc info
   ```
   [root@KONG:stapxx (master)]# ./stap++ samples/lj-gc.sxx -x 19047
   Found exact match for libluajit: /usr/local/openresty/luajit/lib/libluajit-5.1.so.2.1.0
   Start tracing 19047 (/usr/local/openresty/nginx/sbin/nginx)
   Total GC count: 194477462 bytes

   [root@KONG:stapxx (master)]# ./stap++ samples/lj-gc.sxx -x 19047
   Found exact match for libluajit: /usr/local/openresty/luajit/lib/libluajit-5.1.so.2.1.0
   Start tracing 19047 (/usr/local/openresty/nginx/sbin/nginx)
   Total GC count: 247561267 bytes

   [root@KONG:stapxx (master)]# ./stap++ samples/lj-gc.sxx -x 19047
   Found exact match for libluajit: /usr/local/openresty/luajit/lib/libluajit-5.1.so.2.1.0
   Start tracing 19047 (/usr/local/openresty/nginx/sbin/nginx)
   Total GC count: 314516362 bytes

   [root@KONG:stapxx (master)]# ./stap++ samples/lj-gc.sxx -x 19047
   Found exact match for libluajit: /usr/local/openresty/luajit/lib/libluajit-5.1.so.2.1.0
   Start tracing 19047 (/usr/local/openresty/nginx/sbin/nginx)
   Total GC count: 545262046 bytes

   [root@KONG:stapxx (master)]# ./stap++ samples/lj-gc.sxx -x 19047
   Found exact match for libluajit: /usr/local/openresty/luajit/lib/libluajit-5.1.so.2.1.0
   Start tracing 19047 (/usr/local/openresty/nginx/sbin/nginx)
   Total GC count: 601771312 bytes

   [root@KONG:stapxx (master)]# ./stap++ samples/lj-gc.sxx -x 19047
   Found exact match for libluajit: /usr/local/openresty/luajit/lib/libluajit-5.1.so.2.1.0
   Start tracing 19047 (/usr/local/openresty/nginx/sbin/nginx)
   Total GC count: 678657202 bytes
   ```

### Additional Details & Logs

- Kong version
  `0.12.3`
- Kong debug-level startup logs
  `noting useful info`
- Kong error logs
  `noting useful info`
- Kong configuration
  `default, use no any custom template `
- Operating system
  `CentOS Linux release 7.4.1708 (Core) 64 BIT`

**Comment by p0pr0ck5** [comment]
@ms2008 can you verify if this behavior is observed in the latest release of Kong (1.0)? I was not able to reproduce this, but would like to get some feedback if possible.

**Comment by ms2008** [comment]
@p0pr0ck5  I gave it a try, and no this issue found in the Kong 1.0.

**Comment by ms2008** [comment]
The memory usage stats dumped by lj-gc-objs tool attached below, it seems a FFI problem?
```
[root@KONG:stapxx (master)]# ./stap++ samples/lj-gc-objs.sxx -x 10331                           
Found exact match for libluajit: /usr/local/openresty/luajit/lib/libluajit-5.1.so.2.1.0
Start tracing 10331 (/usr/local/openresty/nginx/sbin/nginx)

main machine code area size: 262144 bytes
C callback machine code size: 4096 bytes
GC total size: 661372355 bytes
GC state: sweep

17586 table objects: max=49184, avg=128, min=32, sum=2263784 (in bytes)
8712 cdata objects: max=12000040, avg=75403, min=12, sum=656913172 (in bytes)
7246 string objects: max=3835, avg=38, min=18, sum=276498 (in bytes)
5850 function objects: max=144, avg=28, min=20, sum=168552 (in bytes)
3960 upvalue objects: max=24, avg=24, min=24, sum=95040 (in bytes)
2605 proto objects: max=4102, avg=308, min=74, sum=802774 (in bytes)
399 thread objects: max=832, avg=742, min=424, sum=296160 (in bytes)
352 trace objects: max=8820, avg=1126, min=160, sum=396568 (in bytes)
26 userdata objects: max=28928, avg=1631, min=32, sum=42431 (in bytes)
JIT state size: 27648 bytes
global state tmpbuf size: 2048 bytes
C type state size: 16856 bytes

My GC walker detected for total 661372355 bytes.
15517 microseconds elapsed in the probe handler.
WARNING: Number of errors: 0, skipped probes: 7
```

**Comment by ms2008** [comment]
Here is a flame-graphs sampled by sample-bt-leaks:

![mem_leak](https://user-images.githubusercontent.com/3797918/51378294-ed566180-1b47-11e9-8cdc-7bccda24f038.png)

**Comment by ms2008** [comment]
Oh, I overlooked something, which is that the `db_cache_ttl` should be configured as 0 and must have multiple worker configured. 

This issue actually caused by plugin configuration cache missed from L1. Then triggered a shm lookup of L2, it always returns a new Lua table. We can ensure this by logging the conf's memory address in bot-detection's handler:

```lua
local cache = ua_caches[conf]
ngx.log(ngx.ERR, "cache: " .. tostring(cache), "|", "conf: " .. tostring(conf))
if not cache then
  cache = lrucache.new(UA_CACHE_SIZE)
  ua_caches[conf] = cache
end
```

![nginx-error-log](https://user-images.githubusercontent.com/3797918/51474243-a079c180-1db9-11e9-910f-2550779a921e.png)

To be precise, there is a bug in mlcache.lua bundled by Kong 0.12.3:

```lua
local function shmlru_get(self, key)
    local v, err = self.dict:get(key)
    if err then
        return nil, "could not read from lua_shared_dict: " .. err
    end

    if v ~= nil then
        local str_serialized, value_type, at, ttl = unmarshallers.shm_value(v)

        local remaining_ttl = ttl - (now() - at)

        -- value_type of 0 is a nil entry
        if value_type == 0 then
            return set_lru(self, key, CACHE_MISS_SENTINEL_LRU, remaining_ttl)
        end

        local value, err = unmarshallers[value_type](str_serialized)
        if err then
            return nil, "could not deserialize value after lua_shared_dict " ..
                        "retrieval: " .. err
        end

        return set_lru(self, key, value, remaining_ttl)
    end
end
```

When `db_cache_ttl` configured as 0, the remaining TTL returns a negative TTL, it would never cached in L1.

**Comment by p0pr0ck5** [comment]
@thibaultcha, do you know if the version of lua-resty-mlcache bundled with recent releases of Kong exhibits the same behavior?

**Comment by thibaultcha** [comment]
@ms2008 Good catch!

@p0pr0ck5 This behavior disappeared when mlcache starting becoming an independent project, and received [this commit](https://github.com/thibaultcha/lua-resty-mlcache/commit/baf747b1a0f3a8f6c447534abeb8e2f61b8278f4), in which we introduced a branch to zero-out the `remaining_ttl` instead of passing it to lua-resty-lrucache as negative, see the commit message:

> More importantly, this fixes an issue with the shm -> LRU setting when
computing the remaining TTL. When the TTL is `0` (indefinite as per this
library's convention), we used to set the value in the LRU with a
negative TTL (`0 - (now() - at)`), which resulted in the value actually
not being cached in the LRU because the TTL would be negative (already
expired).

So this should be fixed starting with Kong 0.13.1, according to the first time mlcache was added as a LuaRocks library ([source](https://github.com/Kong/kong/blob/0.13.1/kong-0.13.1-0.rockspec#L34)).
